### PR TITLE
ENG-304: CLI is capable of updating itself in place

### PR DESCRIPTION
**Pull request by andresatvery**
Why:

We want to ensure we can upgrade our CLI by getting the latest release from GitHub.

How:

- adding the `upgrade` command, which lets us download the latest release on GitHub
- download the tar.gz in memory and place the file on cache, respecting XDG

### CODE REVIEW

File: src/api/mod.rs [hunk 1]
```diff
@@ -19,13 +20,16 @@ pub struct Command<T: StructOpt> {
 pub enum ApiCommand {
     SigningKeys(signing_keys::SigningKeysCommand),
     Users(users::UsersCommand),
+    #[structopt(flatten)]
+    Update(update::UpdateCommand),
```
Comment (danielspofford) [inline src/api/mod.rs:24]: If this impacts the order the commands are listed in `--help` lets keep it alphabetical.

Command should also be `Upgrade` not `Update` - will have to change that in a couple spots across the PR but only mentioning it here.

File: Cargo.toml [hunk 2]
```diff
@@ -13,3 +13,10 @@ snafu = "0.7"
 structopt = "0.3"
 tokio = { version = "1.4", features = ["full"] }
 tower = {version = "0.4"}
+directories = "4.0.1"
+reqwest = {version = "0.11", default-features = false, features = ["json", "rustls-tls", "stream"]}
+serde = "1.0.144"
+indicatif = "0.17.0"
```
Comment (danielspofford) [inline Cargo.toml:19]: Ha neat - look forward to testing this.

**Review by danielspofford — COMMENTED** [review]
Just glanced at the code and made a couple comments.

Will give it a spin in practice tomorrow. Exciting!

File: Cargo.toml (hunk 2 above)
Comment (andresatvery) [reply Cargo.toml:19]: I know that we didn't wanted to have progress bars in this iteration, but they are easy to implement and they are much better than just waiting for it to upgrade ^^'

File: Cargo.toml [hunk 3]
```diff
@@ -13,3 +13,10 @@ snafu = "0.7"
...
+futures-util = "0.3.24"
+flate2 = "1.0.24"
+tar = "0.4.38"
```
Comment (danielspofford) [inline Cargo.toml:22]: Can we re-add a newline to the end of this file please? Most editors have a setting one can set that ensures on save files end with a newline, may be worth investigating that.

**Review by danielspofford — CHANGES_REQUESTED** [review]
Let's also update the commit to `ENG-304: add upgrade command` per https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?h=v2.36.1#n181.

I know that is a somewhat petty and tricky one to follow, but we will try our best to be consistent in doing so - appreciate it <3!

**Comment by danielspofford** [comment]
Approved, good to merge. I will test this in practice with release candidate tags (prior to publishing an actual next release).

**Pull request merged**